hub/core/pyffmpeg/pyffmpeg.c

```diff
@@ -0,0 +1,234 @@
+#include <libavcodec/avcodec.h>
+#include <libavformat/avformat.h>
+#include <libswscale/swscale.h>
+#include <stdio.h>
+#include <stdarg.h>
+#include <string.h>
+#include <inttypes.h>
+
+static void logging(const char *fmt, ...);
+static int decode_video_packet(AVPacket *pPacket, AVCodecContext *pCodecContext, AVFrame *pFrame, unsigned char **decompressed, struct SwsContext **sws_context, int *bufpos);
+int readFunc(void *opaque, uint8_t *buf, int buf_size);
+
+struct buffer_data
+{
+    uint8_t *ptr;
+    size_t size;
+};
+
+int getVideoShape(unsigned char *file, int size, int ioBufferSize, int *shape, int isBytes)
+{
+    av_log_set_level(AV_LOG_QUIET);
+    AVFormatContext *pFormatContext = NULL;
+    AVIOContext *pioContext = NULL;
+    unsigned char *ioBuffer;
+    pFormatContext = avformat_alloc_context();
+    struct buffer_data bd = {0};
+
+    int ret = 0;
+
+    if (isBytes == 1)
+    {
+        bd.ptr = file;
+        bd.size = size;
+        ioBuffer = av_malloc(ioBufferSize);
+        pioContext = avio_alloc_context(ioBuffer, ioBufferSize, 0, &bd, &readFunc, NULL, NULL);
+        pFormatContext->pb = pioContext;
+        ret = avformat_open_input(&pFormatContext, NULL, NULL, NULL);
+    }
+    else
+    {
+        ret = avformat_open_input(&pFormatContext, (const char *)file, NULL, NULL);
+    }
+
+    avformat_find_stream_info(pFormatContext, NULL);
+
+    for (int i = 0; i < pFormatContext->nb_streams; i++)
+    {
+
+        AVCodecParameters *pLocalCodecParameters = NULL;
+        pLocalCodecParameters = pFormatContext->streams[i]->codecpar;
+
+        if (pLocalCodecParameters->codec_type == AVMEDIA_TYPE_VIDEO)
+        {
+            float fps = (float)pFormatContext->streams[i]->avg_frame_rate.num / (float)pFormatContext->streams[i]->avg_frame_rate.den;
+            float timebase = (float)pFormatContext->streams[i]->time_base.num / (float)pFormatContext->streams[i]->time_base.den;
+            float duration = (float)pFormatContext->streams[i]->duration * timebase;
+            if (duration < 0)
+            {
+                duration = (float)pFormatContext->duration / AV_TIME_BASE;
+            }
+            int n_frames = (int)duration * (int)fps;
+            int width = pLocalCodecParameters->width;
+            int height = pLocalCodecParameters->height;
+            shape[0] = n_frames;
+            shape[1] = height;
+            shape[2] = width;
+            break;
+        }
+    }
+    avformat_close_input(&pFormatContext);
+    if (isBytes == 1)
+    {
+        if (pioContext)
+        {
+            av_freep(&pioContext->buffer);
+            av_freep(&pioContext);
+        }
+    }
+    return 0;
+}
+
+int decompressVideo(unsigned char *file, int size, int ioBufferSize, unsigned char *decompressed, int isBytes, int nbytes)
+{
+    av_log_set_level(AV_LOG_QUIET);
+    AVFormatContext *pFormatContext = NULL;
```

Comment:
Can we allocate context once and store it in python object? It seems redundant to me to do it every time. Also prone to memory leak of something goes wrong.

Reply:
I don't think so because we will have a pointer to a pointer and when the underlying object gets freed we will point to garbage.